File: partie2/python/multi_agent.py
import sys
import math
import random
import csv

# ------------------------------
# Paramètres du modèle (fixes)
# ------------------------------
GRID_SIZE = 300
N_AGENTS = 20000
DAYS = 730
BETA = 0.5   # force d'infection (dans exp(-BETA * Ni))

# ------------------------------
# PRNG simple et reproductible (Xorshift64)
# ------------------------------
class Xorshift:
    def __init__(self, seed):
        self.state = seed ^ 0xFFFFFFFFFFFFFFFF  # bon mélange initial
        if self.state == 0:
            self.state = 1

    def random(self):
        """Retourne float dans [0, 1)"""
        self.state ^= (self.state << 13) & 0xFFFFFFFFFFFFFFFF
        self.state ^= (self.state >> 7)  & 0xFFFFFFFFFFFFFFFF
        self.state ^= (self.state << 17) & 0xFFFFFFFFFFFFFFFF
        return self.state / 0xFFFFFFFFFFFFFFFF

# ------------------------------
# Tirage exponentiel (durées dE, dI, dR)
# ------------------------------
def neg_exp(mean, rng):
    # -mean * log(1 - U) avec U ~ Uniform(0,1)
    u = rng.random()
    while u == 0:  # évite log(0)
        u = rng.random()
    return math.ceil(-mean * math.log(u))  # ceil pour avoir entier >=1

# ------------------------------
# Classe Agent
# ------------------------------
class Agent:
    def __init__(self, status, de, di, dr, x, y):
        self.status = status          # 0=S, 1=E, 2=I, 3=R
        self.time_in_status = 0
        self.dE = de
        self.dI = di
        self.dR = dr
        self.x = x
        self.y = y

# ------------------------------
# Fonction principale
# ------------------------------
def main():
    if len(sys.argv) != 3:
        print("Usage: python multi_agent.py <rep_id> <output_csv>")
        sys.exit(1)

    rep_id = int(sys.argv[1])
    output_file = sys.argv[2]

    # Seed différent par réplication
    rng = Xorshift(rep_id * 123456789 + 987654321)

    # Création grille (liste de listes d'agents)
    grid = [[[] for _ in range(GRID_SIZE)] for _ in range(GRID_SIZE)]

    agents = []
    for i in range(N_AGENTS):
        if i < 20:
            status = 2  # I
        else:
            status = 0  # S

        dE = neg_exp(3,   rng)
        dI = neg_exp(7,   rng)
        dR = neg_exp(365, rng)

        x = int(rng.random() * GRID_SIZE)
        y = int(rng.random() * GRID_SIZE)

        agent = Agent(status, dE, dI, dR, x, y)
        grid[x][y].append(agent)
        agents.append(agent)

    # Fichier de sortie
    with open(output_file, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerow(['time', 'S', 'E', 'I', 'R'])

        for day in range(DAYS + 1):
            # Comptage
            counts = [0] * 4
            for ag in agents:
                counts[ag.status] += 1
            writer.writerow([day] + counts)

            if day == DAYS:
                break

            # Mélange aléatoire des agents → asynchrone
            random.shuffle(agents)   # ← important !

            for agent in agents:
                # 1. Déplacement aléatoire (téléportation complète)
                new_x = int(rng.random() * GRID_SIZE)
                new_y = int(rng.random() * GRID_SIZE)

                grid[agent.x][agent.y].remove(agent)
                agent.x = new_x
                agent.y = new_y
                grid[new_x][new_y].append(agent)

                # 2. Mise à jour du statut (toujours après déplacement)
                agent.time_in_status += 1

                if agent.status == 1:  # E → I ?
                    if agent.time_in_status >= agent.dE:
                        agent.status = 2
                        agent.time_in_status = 0

                elif agent.status == 2:  # I → R ?
                    if agent.time_in_status >= agent.dI:
                        agent.status = 3
                        agent.time_in_status = 0

                elif agent.status == 3:  # R → S ?
                    if agent.time_in_status >= agent.dR:
                        agent.status = 0
                        agent.time_in_status = 0

                elif agent.status == 0:  # S → E ?
                    ni = 0
                    for dx in [-1, 0, 1]:
                        for dy in [-1, 0, 1]:
                            nx = (agent.x + dx) % GRID_SIZE
                            ny = (agent.y + dy) % GRID_SIZE
                            ni += len(grid[nx][ny])   # tous les agents sur la cellule voisine

                    if ni > 0:   # optimisation : pas besoin si 0
                        p = 1 - math.exp(-BETA * ni)
                        if rng.random() < p:
                            agent.status = 1
                            agent.time_in_status = 0

    print(f"Réplication {rep_id} terminée → {output_file}")

if __name__ == "__main__":
    main()

File: partie2/c/multi_agent.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

#define GRID_SIZE 300
#define N_AGENTS 20000
#define DAYS 730
#define BETA 0.5

typedef struct {
    int status;           // 0=S, 1=E, 2=I, 3=R
    int time_in_status;
    int dE, dI, dR;
    int x, y;
} Agent;

unsigned long long xorshift_state;

double xorshift_rand() {
    xorshift_state ^= xorshift_state << 13;
    xorshift_state ^= xorshift_state >> 7;
    xorshift_state ^= xorshift_state << 17;
    return (double)xorshift_state / (double)(0xFFFFFFFFFFFFFFFFULL);
}

void xorshift_seed(unsigned long long seed) {
    xorshift_state = seed;
    if (xorshift_state == 0) xorshift_state = 1;
}

int neg_exp(double mean) {
    double u = xorshift_rand();
    if (u <= 1e-10 || u >= 1.0 - 1e-10) u = 0.5;
    return (int)ceil(-mean * log(u));
}

int *cell_agents[GRID_SIZE][GRID_SIZE];
int cell_counts[GRID_SIZE][GRID_SIZE];

int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s <rep_id> <output.csv>\n", argv[0]);
        return 1;
    }

    int rep_id = atoi(argv[1]);
    char *output_file = argv[2];

    xorshift_seed((unsigned long long)rep_id * 123456789ULL + 987654321ULL);

    memset(cell_counts, 0, sizeof(cell_counts));
    memset(cell_agents, 0, sizeof(cell_agents));

    Agent *agents = malloc(N_AGENTS * sizeof(Agent));
    if (!agents) {
        perror("malloc agents");
        return 1;
    }

    for (int i = 0; i < N_AGENTS; i++) {
        agents[i].status = (i < 20) ? 2 : 0;
        agents[i].time_in_status = 0;
        agents[i].dE = neg_exp(3.0);
        agents[i].dI = neg_exp(7.0);
        agents[i].dR = neg_exp(365.0);
        agents[i].x = (int)(xorshift_rand() * GRID_SIZE);
        agents[i].y = (int)(xorshift_rand() * GRID_SIZE);

        int x = agents[i].x, y = agents[i].y;
        int *new_list = realloc(cell_agents[x][y], (cell_counts[x][y] + 1) * sizeof(int));
        if (!new_list) {
            fprintf(stderr, "realloc failed\n");
            free(agents);
            return 1;
        }
        cell_agents[x][y] = new_list;
        cell_agents[x][y][cell_counts[x][y]] = i;
        cell_counts[x][y]++;
    }

    FILE *fp = fopen(output_file, "w");
    if (!fp) {
        perror("fopen");
        free(agents);
        return 1;
    }

    fprintf(fp, "time,S,E,I,R\n");

    int order[N_AGENTS];
    for (int i = 0; i < N_AGENTS; i++) order[i] = i;

    for (int day = 0; day <= DAYS; day++) {
        int counts[4] = {0,0,0,0};
        for (int i = 0; i < N_AGENTS; i++)
            counts[agents[i].status]++;

        fprintf(fp, "%d,%d,%d,%d,%d\n", day, counts[0], counts[1], counts[2], counts[3]);

        if (day == DAYS) break;

        // Fisher-Yates shuffle
        for (int i = N_AGENTS - 1; i > 0; i--) {
            int j = (int)(xorshift_rand() * (i + 1));
            int temp = order[i];
            order[i] = order[j];
            order[j] = temp;
        }

        for (int idx = 0; idx < N_AGENTS; idx++) {
            int i = order[idx];
            Agent *a = &agents[i];

            // Déplacement
            int old_x = a->x, old_y = a->y;
            int new_x = (int)(xorshift_rand() * GRID_SIZE);
            int new_y = (int)(xorshift_rand() * GRID_SIZE);

            // Retirer de l'ancienne cellule
            int *list = cell_agents[old_x][old_y];
            int pos = -1;
            for (int k = 0; k < cell_counts[old_x][old_y]; k++) {
                if (list[k] == i) {
                    pos = k;
                    break;
                }
            }
            if (pos >= 0) {
                list[pos] = list[cell_counts[old_x][old_y] - 1];
                cell_counts[old_x][old_y]--;
            }

            // Ajouter à la nouvelle
            int *new_list = realloc(cell_agents[new_x][new_y], (cell_counts[new_x][new_y] + 1) * sizeof(int));
            if (!new_list) {
                fprintf(stderr, "realloc failed\n");
                fclose(fp);
                free(agents);
                return 1;
            }
            cell_agents[new_x][new_y] = new_list;
            cell_agents[new_x][new_y][cell_counts[new_x][new_y]] = i;
            cell_counts[new_x][new_y]++;

            a->x = new_x;
            a->y = new_y;

            // Mise à jour statut
            a->time_in_status++;

            if (a->status == 1 && a->time_in_status >= a->dE) {
                a->status = 2;
                a->time_in_status = 0;
            } else if (a->status == 2 && a->time_in_status >= a->dI) {
                a->status = 3;
                a->time_in_status = 0;
            } else if (a->status == 3 && a->time_in_status >= a->dR) {
                a->status = 0;
                a->time_in_status = 0;
            } else if (a->status == 0) {
                int ni = 0;
                for (int dx = -1; dx <= 1; dx++) {
                    for (int dy = -1; dy <= 1; dy++) {
                        int nx = (a->x + dx + GRID_SIZE) % GRID_SIZE;
                        int ny = (a->y + dy + GRID_SIZE) % GRID_SIZE;
                        ni += cell_counts[nx][ny];
                    }
                }
                if (ni > 0) {
                    double p = 1.0 - exp(-BETA * (double)ni);
                    if (xorshift_rand() < p) {
                        a->status = 1;
                        a->time_in_status = 0;
                    }
                }
            }
        }
    }

    fclose(fp);

    // Libération mémoire
    for (int i = 0; i < GRID_SIZE; i++) {
        for (int j = 0; j < GRID_SIZE; j++) {
            free(cell_agents[i][j]);
        }
    }
    free(agents);

    printf("Réplication %d terminée → %s\n", rep_id, output_file);
    return 0;
}
